### DISCUSSION TITLE
type_annotation_map doesn't seem to apply to `column_property`

### DISCUSSION THREAD

**Discussion by tamird** [sqlalchemy/sqlalchemy#11548] [Usage Questions] [answered]
### Describe the bug

I have a materialized column in my model that I want to replace with a `column_property`. I wrote this:
```python
class ...(....):
   ...
    _validation_status_nhh: Mapped[CumValidationNHH] = mapped_column(name="validation_status_nhh")

    @hybrid_property
    def validation_status_nhh(self) -> CumValidationNHH:
        assert self._validation_status_nhh == self.computed_validation_status_nhh # doesn't assert
        assert isinstance(self._validation_status_nhh, CumValidationNHH) # doesn't assert
        assert isinstance(
            self.computed_validation_status_nhh,
            CumValidationNHH,
        ), repr(self.computed_validation_status_nhh) # this asserts, self.computed_validation_status_nhh is a str
        return self._validation_status_nhh

    @validation_status_nhh.inplace.expression
    def _validation_status_nhh_expression(cls) -> coalesce[CumValidationNHH]:
        return _computed_validation_status_nhh(cls.id, cls.calculated_at_utc)

    computed_validation_status_nhh: Mapped[CumValidationNHH] = column_property(
        _computed_validation_status_nhh(id, calculated_at_utc),
    )
```

Apart from this everything works correctly - I even confirmed that `computed_validation_status_nhh` and `validation_status_nhh` are equal for all objects in my database, so I'm certain the implementation of `computed_validation_status_nhh` is correct.

### SQLAlchemy Version in Use

2.0.31

### DBAPI (i.e. the database driver)

asyncpg

### Database Vendor and Major Version

postgres 16

### Python Version

3.12

### Operating system

Linux

### To Reproduce

I don't have a self contained example, but there are no tests covering this either. It should be easy to add a test that combines type_annotation_map and column_property.

### Error

N/A

**Comment by CaselIT** [comment]
Hi,

I don't understand what you mean.
The `type_annotation_map` is used to infer the db type from the python type while column_property is usually used instead to define an expression using existing column.

There is no runnable script that can clarify do moving to discussion

**Reply by tamird** [reply]
> The type_annotation_map is used to infer the db type from the python type while column_property is usually used instead to define an expression using existing column.

Exactly - I'm trying to use `column_property` to return a string from the database but I want the python type to be a `StrEnum`defined in my application code. Using `type_annotation_map` for this string->enum inference works for `mapped_column` but not for `column_property`; the value comes back as a plain `str` rather than having been passed through my `TypeAdapter`.

Does that clarify?

**Reply by tamird** [reply]
@CaselIT what can I do to help move this forward?

**Reply by CaselIT** [reply]
As far as I know that's not supported.

When you use a mapped_column its sql type is set an Enum that manages the conversion to enum.
column property is an ORM concept and I'm not sure if it can use the same logic

@zzzeek do you have suggestions here?

**Comment by zzzeek** [comment]
I have a guess what you might be trying to do but it would help if the code could maybe slightly indicate this.

Your code has one column_property:

```
  computed_validation_status_nhh: Mapped[CumValidationNHH] = column_property(
        _computed_validation_status_nhh(id, calculated_at_utc),
    )
```

ive copy-pasted the code into an editor so I can use search to make sure I'm not missing identifiers.   I have no indication what `_computed_validation_status_nhh`, `id`, or `calculated_at_utc` refer to.   I see a `mapped_column()` and a hybrid property that I assume are related, but these use different names, so I dont really know what the problem is.  

> so I'm certain the implementation of computed_validation_status_nhh is correct.

I can't help because I dont see any such implementation.

Based on what I'm guessing you might be trying to do, overall, there is no interaction with the type_annotation_map and column_property() at runtime in order to determine a SQL type, because SQL expressions given to column_property() already have a SQL type.  There is no place where these two things would interact.

**Reply by zzzeek** [reply, marked as answer]
if....you want your column_property() which normally returns a string to do a StrEnum, use [type_coerce](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.type_coerce)

```
computed_validation_status_nhh: Mapped[CumValidationNHH] = column_property(
         type_coerce(
               _computed_validation_status_nhh(id, calculated_at_utc),
              Enum(MyStrEnum),
         )
    )
```

**Reply by tamird** [reply]
Yeah so my `column_property` is joining to another table and selecting a property from that model which is a `mapped_column` of the `StrEnum` type. That's where the interaction with the type_annotation_map would come from (in my mind, at least).

**Reply by zzzeek** [reply]
the type_annotation_map is used to give a SQL type to a `mapped_column()` that has **no** SQL type to start with.  if you give your `mapped_column()` a SQL type, like `mapped_column(Integer)`, type_annotation_map is not used.   the SQL expression you pass to `column_property()` similarly already has a SQL type.  if you wanted it to be a different type, you need to alter the expression to set up the means by which you want that type to be converted, such as `cast()` or `type_coerce()`.    There is not just one choice and type_annotation_map is not going to invent an arbitrary coercion implicitly.

**Reply by tamird** [reply]
Makes sense. Perhaps this could be clarified in the documentation of `type_annotation_map` - or perhaps it is already clear and I didn't notice.

**Reply by zzzeek** [reply]
The section at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapped-column-derives-the-datatype-and-nullability-from-the-mapped-annotation talks about mapped_column() and indicates "If the mapped_column() construct indicates an explicit type as passed to the mapped_column.__type argument, then the given Python type is disregarded."   so....it is not talking about column_property() at all in the first place.   not sure how to bring this up

**Reply by tamird** [reply]
I think the problem here is that `column_property` is permitted to return non-SQL types. Looking at [its signature](https://github.com/sqlalchemy/sqlalchemy/blob/8f47774de91b18e8b9a21d39662bffd239785b4e/lib/sqlalchemy/orm/_orm_constructors.py#L438-L455) you can see that it is generic on `_T` which is permitted to be any type.

Do you think it'd make sense to narrow this to a SQL type to prevent this mistake?

**Reply by zzzeek** [reply]
im not following.    both the type passed to it and the return type are typed as SQL expressions

**Reply by tamird** [reply]
They are SQL expressions but they are permitted to return *any* type, which in my case was an application-defined `StrEnum`. I'm saying the type should be restricted to a DB-native type like `TEXT`, `JSONB`, etc. This restriction would have prevented me from shooting myself in the foot as I did here.

**Reply by zzzeek** [reply]
they return a SQL expression with Python type `_T`, that is, the expression given on the inside if it were "str" means the outer expression is also "str".     When we say "type" here we are talking about Python types.

as far as limiting the SQL `TypeEngine` to some specific subset of types, that's already accomplished by the `_T`.  I dont follow what "DB native" has to do with anything here.   

if you can show a complete, self contained example of what you are talking about, including exact code that you feel should not be accepted by the type checker, that would be very helpful, thanks

**Reply by tamird** [reply]
It's quite hard to show you a complete example, but look at the snippet I showed above:
```python
    computed_validation_status_nhh: Mapped[CumValidationNHH] = column_property(
        _computed_validation_status_nhh(id, calculated_at_utc),
    )
```

given that `CumValidationNHH` is a `StrEnum` I've defined in my application it is obviously never going to be returned from a database query. At a high level I'm saying that the type checker should reject this.

**Reply by zzzeek** [reply]
you're asking me to change the library in a way that I dont understand.   Please provide an example

**Reply by zzzeek** [reply]
> given that CumValidationNHH is a StrEnum I've defined in my application

why not?   We have an Enum type and from your mapping it looks like that type would be used here, you'd get your StrEnum back

if you look at your example, there is **no** context given for any of those types, and there's tons of asserts.    Just put the imports in, resolve the symbols so I can see what we're talking about.    As I mentioned previously, ". I have no indication what _computed_validation_status_nhh, id, or calculated_at_utc refer to"  these are simple things you can illustrate.  It's important that you're able to look at your example above in the way that someone else sees it, and explain the various elements going on.    if it's "too complicated", that points to maybe the thing you're asking for doesnt actually make sense if you were to think it through, I really dont know.

**Reply by tamird** [reply]
Ack. A detail worth mentioning here is that we don't use native enums - the enum is stored as a string.

**Reply by zzzeek** [reply]
the SQLAlchemy [Enum](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum) type supports a string column translated back to the `enum.Enum` of your choice.  is that what's missing here?

**Reply by tamird** [reply]
Why would that make a difference? Reading the [documentation](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#altering-the-configuration-of-the-default-enum) it seems that the SQLAlchemy `Enum` type provides little more than a default in `type_annotation_map`. Given that `column_property` doesn't consult `type_annotation_map`, it seems that using `Enum` won't change the outcome.

I'll try it, but I don't see how it'd help.

EDIT: I've made the following change in `type_annotation_map`:
```diff
-        CumValidationHH: TEXT(),
-        CumValidationNHH: TEXT(),
+        enum.StrEnum: sa.Enum(enum.StrEnum, native_enum=False),
```
and indeed I see the same issue.

**Reply by zzzeek** [reply]
I would register your actual StrEnum subclass in the map with the datatypes you are looking to be matched to it.   We probably aren't looking at arbitrary subclasses of StrEnum right now.

**Reply by tamird** [reply]
If I've understood you correctly:
```diff
diff --git a/lib/common/common/db/model_base.py b/lib/common/common/db/model_base.py
index ba77fe837e..974073f61c 100644
--- a/lib/common/common/db/model_base.py
+++ b/lib/common/common/db/model_base.py
@@ -17,6 +17,7 @@ from sqlalchemy import (
     CompoundSelect,
     Date,
     DateTime,
+    Enum,
     Select,
     inspect,
     select,
@@ -265,8 +266,8 @@ TYPE_ANNOTATION_MAP: Mapping[Any, Any] = {
         CopyIcon: TEXT(),
         CopyTextContentKey: TEXT(),
         COSGainType: TEXT(),
-        CumValidationHH: TEXT(),
-        CumValidationNHH: TEXT(),
+        CumValidationHH: Enum(CumValidationHH, native_enum=False),
+        CumValidationNHH: Enum(CumValidationHH, native_enum=False),
         Currency: TEXT(),
         CustomerPIIPropertyType: TEXT(),
         CustomerPreferencesType: TEXT(),
diff --git a/lib/common/common/module/consumption/db/models.py b/lib/common/common/module/consumption/db/models.py
index 6b0d38c93a..745046bf6a 100644
--- a/lib/common/common/module/consumption/db/models.py
+++ b/lib/common/common/module/consumption/db/models.py
@@ -1214,7 +1214,7 @@ class Consumption_Cumulative_Parent(DbBaseModel, TimeStampMixin):
         computed_validation_status_nhh = self.computed_validation_status_nhh
         if not isinstance(computed_validation_status_nhh, CumValidationNHH):
             # See https://github.com/sqlalchemy/sqlalchemy/discussions/11548.
-            computed_validation_status_nhh = CumValidationNHH(computed_validation_status_nhh)
+            raise ValueError
         return computed_validation_status_nhh
 
     @validation_status_nhh.inplace.expression
@@ -1246,7 +1246,7 @@ class Consumption_Cumulative_Parent(DbBaseModel, TimeStampMixin):
         computed_validation_status_hh = self.computed_validation_status_hh
         if not isinstance(computed_validation_status_hh, CumValidationHH):
             # See https://github.com/sqlalchemy/sqlalchemy/discussions/11548.
-            computed_validation_status_hh = CumValidationHH(computed_validation_status_hh)
+            raise ValueError
         return computed_validation_status_hh
 
     @validation_status_hh.inplace.expression
```
That didn't work - my tests hit the `ValueError`.

See what I wrote above:

> Why would that make a difference? [...]

I continue to not understand.

**Reply by zzzeek** [reply]
Here is a complete example of how to include a `StrEnum` subclass in the type annotation map also demonstrating that no string values are exposed, all outgoing and incoming values in the Python program are handled as members of the `StrEnum`:

```py
from __future__ import annotations

import enum

from sqlalchemy import create_engine
from sqlalchemy import Enum
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session

class MyEnum(enum.StrEnum):
    one = "one"
    two = "two"


class Base(DeclarativeBase):
    type_annotation_map = {
        MyEnum: Enum(MyEnum, native_enum=False),
    }



class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[MyEnum]

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as s:
    s.add(A(data=MyEnum.two))
    s.commit()

with Session(e) as s:
    assert s.scalar(select(A.data)) is MyEnum.two

```

**Reply by tamird** [reply]
Yes, this is 99% of our usage, and it works! Where it breaks down is that if you use a `column_property` that references one of these `Mapped[StrEnum]`s you will get a string, and not an enum.

**Reply by zzzeek** [reply]
you can also map Enums to native_enum=False generically as illustrated in the docs like this

```py
class Base(DeclarativeBase):
    type_annotation_map = {
        enum.Enum: Enum(enum.Enum, native_enum=False),
    }

```

**Reply by tamird** [reply]
I think using your example you can spot the problem by just adding `data_property: Mapped[MyEnum] = column_property(data)` to `A`.

**Reply by zzzeek** [reply]
column_property() will return whatever the type of the SQL expression is.   It's been two weeks and I still have no further information what your actual SQL expression does but if it resolves to that same column's datatype, that's what you'll get.  it's very likely you need to type_coerce() however if you're trying to get a string expression to resolve into an enum as below

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[MyEnum] = mapped_column()

    data_plus_thing: Mapped[MyEnum] = column_property(
        type_coerce("t" + "wo", Enum(MyEnum, native_enum=False))
    )

```

there's a typing issue there - Enum extends `String` and that's making the above not type correctly.  is that the issue?

added #11597 for that issue, seems to only be pyright

**Reply by tamird** [reply]
In your example you're making a column property from a string literal, so it is understandable that it doesn't return the annotated type. What I'm saying is that we get the wrong type even when the column property selects from a mapped_column that has the correct type (via the type annotation map).

I think this complete example illustrates the point:
```python
from __future__ import annotations

import enum

from sqlalchemy import create_engine
from sqlalchemy import Enum
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import mapped_property
from sqlalchemy.orm import Session

class MyEnum(enum.StrEnum):
    one = "one"
    two = "two"


class Base(DeclarativeBase):
    type_annotation_map = {
        MyEnum: Enum(MyEnum, native_enum=False),
    }



class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[MyEnum] = mapped_column()
    data_property: Mapped[MyEnum] = column_property(data)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as s:
    s.add(A(data=MyEnum.two))
    s.commit()

with Session(e) as s:
    assert s.scalar(select(A.data)) is MyEnum.two
    assert s.scalar(select(A.data_property)) is MyEnum.two # boom
```
> there's a typing issue there - `Enum` extends String and that's making the above not type correctly. is that the issue?

Yes, that's certainly an issue highlighted by the code you wrote. But consider my example here - it's even more surprising (to me).

**Reply by zzzeek** [reply]
this doesnt work:

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[MyEnum] = mapped_column()
    data_property: Mapped[MyEnum] = column_property(data)

```

that breaks the program because the data_property is confusing the mapping regarding "data", since you are mapping the column twice. it produces this warning:

```py
/home/classic/dev/sqlalchemy/test3.py:22: SAWarning: On class 'A', Column object 'data' named directly multiple times, only one will be used: data, data_property. Consider using orm.synonym instead
  class A(Base):
```

I need to see a real world example of column_property() that returns this enum.  if it's a SQL function then you need to put the `type_` in that function.   I dont see possibility where the expression inside of `column_property()` can get away without explicitly stating it returns the `Enum(MyEnum)` - this is not going to be propagated from the annotation.

**Reply by tamird** [reply]
I appreciate your continued help, and I know this is frustrating. I'm trying to be mindful of both my employer's proprietary code and your need to parse huge volumes of code.

Adding `type_` to the function did work, and now I understand what you mean - the function can't know the destination type. For reference, here's what produces the expression we use in `column_property`:
```python

def _computed_validation_status(
    validation: type[CumValidationT],
    id: ConsumptionCumParentId | Mapped[ConsumptionCumParentId],
    calculated_at_utc: UtcNaive | Mapped[UtcNaive],
    validation_status: InstrumentedAttribute[CumValidationT],
    *,
    backoffice_validation_causes: Collection[ConsumptionCumulativeCause],
    automatic_validation_causes: Collection[ConsumptionCumulativeCause],
) -> coalesce[CumValidationT]:
    return func.coalesce(
        *(
            func.nullif(
                select(validation_status)
                .where(
                    Consumption_Cumulative_Event_Parent_Mapping.parent_id == id,
                    Consumption_Cumulative_Event_Parent_Mapping.parent_calculated_at_utc == calculated_at_utc,
                    Consumption_Cumulative_Event.id == Consumption_Cumulative_Event_Parent_Mapping.event_id,
                    Consumption_Cumulative_Event.is_from_fuse(),  # pylint: disable=no-value-for-parameter # See https://github.com/pylint-dev/pylint/issues/9757.
                    Consumption_Cumulative_Event.is_to_fuse(),  # pylint: disable=no-value-for-parameter # See https://github.com/pylint-dev/pylint/issues/9757.
                    Consumption_Cumulative_Event.cause.in_(causes),
                )
                .order_by(
                    Consumption_Cumulative_Event.calculated_at_utc.desc(),
                    Consumption_Cumulative_Event.id.desc(),
                )
                .limit(1)
                .correlate_except(Consumption_Cumulative_Event, Consumption_Cumulative_Event_Parent_Mapping)
                .scalar_subquery(),
                validation.U,
            )
            for causes in (backoffice_validation_causes, automatic_validation_causes)
        ),
        validation.U,
        # See https://github.com/sqlalchemy/sqlalchemy/discussions/11548.
        type_=Enum(validation, native_enum=False), # Added per your advice, solves the problem.
    )
```
Perhaps the unsatisfying bit here is that the expression has type `coalesce[Any]`, which I think is caused by the absence of type hints for `func.nullif`.

**Reply by zzzeek** [reply]
there's no need to share employer code for these kinds of issues, whats needed is to distill the pattern down to its essentials and just change all the names of things to A, B, C, etc.   there's a lot of cognitive overhead for someone on the outside to distinguish the difference between names like Consumption_Cumulative_Event, ConsumptionCumulativeCause, Consumption_Cumulative_Event_Parent_Mapping , I have to cut and paste this code into an editor and do a search on each name to make sure I'm not missing a subtle difference in two names.  Similarly for the SQL function just package coalesce and nullif with some dummy input.

with the fix in #11597 pyright types this correctly, this illustrates coalesce + nullif.  nullif is not one of the SQL functions that we have coded / typed in our functions module, we only have a very small list

```py
from __future__ import annotations

import enum

from sqlalchemy import annotations
from sqlalchemy import create_engine
from sqlalchemy import Enum
from sqlalchemy import func
from sqlalchemy import literal
from sqlalchemy import select
from sqlalchemy.orm import annotations
from sqlalchemy.orm import column_property
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session

class MyEnum(enum.StrEnum):
    one = "one"
    two = "two"


class Base(DeclarativeBase):
    type_annotation_map = {
        MyEnum: Enum(MyEnum, native_enum=False),
    }


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)

    data: Mapped[MyEnum] = column_property(
        func.coalesce(
            *(
                func.nullif(
                    select(literal("two")),
                    select(literal(thing)),
                    type_=Enum(MyEnum)
                ) for thing in ["two", "one"]
            )
        )
    )

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as s:
    s.add(A(data=MyEnum.two))
    s.commit()

with Session(e) as s:
    assert s.scalar(select(A.data)) is MyEnum.two

```

**Reply by tamird** [reply]
Indeed, typings for `nullif` would help. I commented [here](https://github.com/sqlalchemy/sqlalchemy/issues/6810#issuecomment-2223622596).

I'd be happy to contribute if mentorship is available.
